File: NowPlaying/Android/src/org/metasyntactic/NowPlayingController.java
//Copyright 2008 Cyrus Najmabadi
//
//Licensed under the Apache License, Version 2.0 (the "License");
//you may not use this file except in compliance with the License.
//You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
//Unless required by applicable law or agreed to in writing, software
//distributed under the License is distributed on an "AS IS" BASIS,
//WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//See the License for the specific language governing permissions and
//limitations under the License.

package org.metasyntactic;

import android.content.Context;
import org.metasyntactic.caches.scores.ScoreType;
import org.metasyntactic.data.*;
import org.metasyntactic.threading.ThreadingUtilities;

import java.util.Date;
import java.util.List;

public class NowPlayingController {
  private final NowPlayingModel model;
  private final Object lock = new Object();

  public NowPlayingController(final Context applicationContext) {
    this.model = new NowPlayingModel(applicationContext);
  }

  public void startup() {
    this.model.startup();
    update();
  }

  public void shutdown() {
    this.model.shutdown();
  }

  private void update() {
    final Runnable runnable = new Runnable() {
      public void run() {
        NowPlayingController.this.model.update();
      }
    };

    ThreadingUtilities.performOnBackgroundThread("Update Model", runnable, this.lock, false/* visible */);
  }

  public String getUserLocation() {
    return this.model.getUserLocation();
  }

  public void setUserLocation(final String userLocation) {
    this.model.setUserLocation(userLocation);
    update();
  }

  public int getSearchDistance() {
    return this.model.getSearchDistance();
  }

  public void setSearchDistance(final int searchDistance) {
    this.model.setSearchDistance(searchDistance);
  }

  public int getSelectedTabIndex() {
    return this.model.getSelectedTabIndex();
  }

  public void setSelectedTabIndex(final int index) {
    this.model.setSelectedTabIndex(index);
  }

  public int getAllMoviesSelectedSortIndex() {
    return this.model.getAllMoviesSelecetedSortIndex();
  }

  public void setAllMoviesSelectedSortIndex(final int index) {
    this.model.setAllMoviesSelectedSortIndex(index);
  }

  public int getAllTheatersSelectedSortIndex() {
    return this.model.getAllTheatersSelectedSortIndex();
  }

  public void setAllTheatersSelectedSortIndex(final int index) {
    this.model.setAllTheatersSelectedSortIndex(index);
  }

  public int getUpcomingMoviesSelectedSortIndex() {
    return this.model.getUpcomingMoviesSelectedSortIndex();
  }

  public void setUpcomingMoviesSelectedSortIndex(final int index) {
    this.model.setUpcomingMoviesSelectedSortIndex(index);
  }

  public List<Movie> getMovies() {
    return this.model.getMovies();
  }

  public List<Theater> getTheaters() {
    return this.model.getTheaters();
  }

  public List<String> getTrailers(final Movie movie) {
    return this.model.getTrailers(movie);
  }

  public List<Review> getReviews(final Movie movie) {
    return this.model.getReviews(movie);
  }

  public String getImdbAddress(final Movie movie) {
    return this.model.getImdbAddress(movie);
  }

  public List<Theater> getTheatersShowingMovie(final Movie movie) {
    return this.model.getTheatersShowingMovie(movie);
  }

  public List<Movie> getMoviesAtTheater(final Theater theater) {
    return this.model.getMoviesAtTheater(theater);
  }

  public List<Performance> getPerformancesForMovieAtTheater(final Movie movie, final Theater theater) {
    return this.model.getPerformancesForMovieAtTheater(movie, theater);
  }

  public ScoreType getScoreType() {
    return this.model.getScoreType();
  }

  public void setScoreType(final ScoreType scoreType) {
    this.model.setScoreType(scoreType);
    update();
  }

  public Score getScore(final Movie movie) {
    return this.model.getScore(movie);
  }

  public byte[] getPoster(final Movie movie) {
    return this.model.getPoster(movie);
  }

  public String getSynopsis(final Movie movie) {
    return this.model.getSynopsis(movie);
  }

  public void prioritizeMovie(final Movie movie) {
    this.model.prioritizeMovie(movie);
  }

  public boolean isAutoUpdateEnabled() {
    return this.model.isAutoUpdateEnabled();
  }

  public void setAutoUpdateEnabled(boolean enabled) {
    this.model.setAutoUpdateEnabled(enabled);
  }

  public Date getSearchDate() {
    return this.model.getSearchDate();
  }

  public void setSearchDate(Date date) {
    this.model.setSearchDate(date);
    update();
  }
}

File: NowPlaying/iPhone/Classes/NowPlayingController.m
// Copyright 2008 Cyrus Najmabadi
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#import "NowPlayingController.h"

#import "Application.h"
#import "ApplicationTabBarController.h"
#import "AlertUtilities.h"
#import "DataProvider.h"
#import "DateUtilities.h"
#import "LocationManager.h"
#import "NowPlayingAppDelegate.h"
#import "NowPlayingModel.h"
#import "ScoreCache.h"
#import "ThreadingUtilities.h"
#import "UpcomingCache.h"
#import "UserLocationCache.h"
#import "Utilities.h"

@interface NowPlayingController()
@property (assign) NowPlayingAppDelegate* appDelegate;
@property (retain) NSLock* determineLocationLock;
@property (retain) LocationManager* locationManager;
@end


@implementation NowPlayingController

@synthesize appDelegate;
@synthesize determineLocationLock;
@synthesize locationManager;

- (void) dealloc {
    self.appDelegate = nil;
    self.determineLocationLock = nil;
    self.locationManager = nil;

    [super dealloc];
}


- (NowPlayingModel*) model {
    return appDelegate.model;
}


- (id) initWithAppDelegate:(NowPlayingAppDelegate*) appDelegate_ {
    if (self = [super init]) {
        self.appDelegate = appDelegate_;
        self.locationManager = [LocationManager managerWithController:self];
    }

    return self;
}


+ (NowPlayingController*) controllerWithAppDelegate:(NowPlayingAppDelegate*) appDelegate {
    return [[[NowPlayingController alloc] initWithAppDelegate:appDelegate] autorelease];
}


- (BOOL) tooSoon:(NSDate*) lastDate {
    if (lastDate == nil) {
        return NO;
    }

    NSDate* now = [NSDate date];

    if (![DateUtilities isSameDay:now date:lastDate]) {
        // different days. we definitely need to refresh
        return NO;
    }

    NSDateComponents* lastDateComponents = [[NSCalendar currentCalendar] components:NSHourCalendarUnit fromDate:lastDate];
    NSDateComponents* nowDateComponents = [[NSCalendar currentCalendar] components:NSHourCalendarUnit fromDate:now];

    // same day, check if they're at least 8 hours apart.
    if (nowDateComponents.hour >= (lastDateComponents.hour + 8)) {
        return NO;
    }

    // it's been less than 8 hours. it's too soon to refresh
    return YES;
}


- (void) spawnDataProviderLookupThread {
    NSAssert([NSThread isMainThread], nil);
    if ([self tooSoon:[self.model.dataProvider lastLookupDate]]) {
        [self onDataProviderUpdateComplete];
    } else {
        [self.model.dataProvider update:self.model.searchDate delegate:self context:nil];
    }
}


- (void) spawnModelUpdateThread {
    NSAssert([NSThread isMainThread], nil);
    [self.model update];
}


- (void) onDataProviderUpdateSuccess:(LookupResult*) lookupResult context:(id) context {
    // Save the results.
    [self.model.dataProvider saveResult:lookupResult];
}


- (void) onDataProviderUpdateFailure:(NSString*) error context:(id) context {
    [self performSelectorOnMainThread:@selector(reportError:) withObject:error waitUntilDone:NO];
}


- (void) reportError:(NSString*) error {
    /*
     UIAlertView* alert = [[[UIAlertView alloc] initWithTitle:nil
     message:NSLocalizedString(@"No information found", nil)
     delegate:nil
     cancelButtonTitle:NSLocalizedString(@"OK", nil)
     otherButtonTitles:nil] autorelease];

     [alert show];
     */
}


- (void) spawnScoresLookupThread {
    NSAssert([NSThread isMainThread], nil);
    [self.model.scoreCache update];
}


- (void) spawnDetermineLocationThread {
    NSAssert([NSThread isMainThread], nil);
    [ThreadingUtilities performSelector:@selector(determineLocationBackgroundEntryPoint)
                               onTarget:self
                   inBackgroundWithGate:determineLocationLock
                                visible:YES];
}


- (void) spawnNetflixThread {
    NSAssert([NSThread isMainThread], nil);
    [self.model.netflixCache update];
}


- (void) spawnUpdateRemainderThreads {
    NSAssert([NSThread isMainThread], nil);
    [self spawnScoresLookupThread];
    [self spawnNetflixThread];
    [self spawnModelUpdateThread];
}


- (void) onDataProviderUpdateComplete {
    NSAssert([NSThread isMainThread], nil);
    [self spawnUpdateRemainderThreads];
}


- (void) start {
    NSAssert([NSThread isMainThread], nil);
    [self spawnDetermineLocationThread];
}


- (void) determineLocationBackgroundEntryPoint {
    NSString* address = self.model.userAddress;
    Location* location = [self.model.userLocationCache downloadUserAddressLocationBackgroundEntryPoint:address];
    
    [self performSelectorOnMainThread:@selector(reportUserLocation:)
                           withObject:location
                        waitUntilDone:NO];
}


- (void) reportUserLocation:(Location*) location {
    NSAssert([NSThread isMainThread], nil);
    if (self.model.userAddress.length > 0 && location == nil) {
        [AlertUtilities showOkAlert:NSLocalizedString(@"Could not find location.", nil)];
    }
    
    [self spawnDataProviderLookupThread];
}


- (void) markDataProviderOutOfDate {
    [self.model.dataProvider markOutOfDate];
}


- (void) setSearchDate:(NSDate*) searchDate {
    if ([DateUtilities isSameDay:searchDate date:self.model.searchDate]) {
        return;
    }

    [appDelegate.tabBarController popNavigationControllersToRoot];
    [self.model setSearchDate:searchDate];

    [self markDataProviderOutOfDate];
    [self spawnDataProviderLookupThread];
}


- (void) setUserAddress:(NSString*) userAddress {
    if ([userAddress isEqual:self.model.userAddress]) {
        return;
    }

    [appDelegate.tabBarController popNavigationControllersToRoot];
    [self.model setUserAddress:userAddress];

    [self markDataProviderOutOfDate];
    [self spawnDetermineLocationThread];

    // Force a refresh so the UI displays this new address.
    [NowPlayingAppDelegate majorRefresh:YES];
}


- (void) setSearchRadius:(NSInteger) radius {
    [self.model setSearchRadius:radius];
}


- (void) setScoreProviderIndex:(NSInteger) index {
    if (index == self.model.scoreProviderIndex) {
        return;
    }

    [self.model setScoreProviderIndex:index];
    [self spawnScoresLookupThread];
}


- (void) setAutoUpdateLocation:(BOOL) value {
    [self.model setAutoUpdateLocation:value];
    [locationManager autoUpdateLocation];
}


- (void) setNetflixEnabled:(BOOL) value {
    [self.model setNetflixEnabled:value];
    [appDelegate.tabBarController resetTabs:YES];
    [NowPlayingAppDelegate majorRefresh:YES];
    [self spawnNetflixThread];
}


- (void) setNetflixKey:(NSString*) key secret:(NSString*) secret userId:(NSString*) userId {
    [self.model setNetflixKey:key secret:secret userId:userId];
    [self spawnNetflixThread];
}

@end